### PR TITLE
[LoRaWAN] Improve PHY behaviour, update beginABP, bugfixes

### PR DESCRIPTION
**Pull request by StevenCellist**
Until TTS v3.30.0 (current version), the ABP keys are displayed incorrectly. With the upcoming TTS v3.30.1, the ABP keys are now displayed exactly as per spec. The `beginABP()` function from this PR onwards uses the exact same order of key arguments with exactly the same names. This also makes it non-obvious how to use LW v1.0.x session which are still not properly supported, as the F/SNwkSIntKeys are now required instead of NULL-default. (They can still be set to NULL.)

* [LoRaWAN] Change beginABP arguments to match official key names & TTN console
* [CI] Update flag for changed LoRaWAN key name

If running LoRaWAN mixed with another protocol sharing the same radio, the physical layer was not reconfigured completely for each uplink. Also, as reminded in Discussions recently, there was a possibility to overrule the Tx power without the LW stack knowing about this. To prevent this, the LW stack now completely configures the PHY layer for each up/downlink.
To accomodate this change and prevent useless configuration of Tx power upon receiving ADR requests or user configuration, the checks for the module's Tx power range in the respective module drivers is separated from their `setOutputPower()` functions into `checkOutputPower()`.

* [LoRaWAN] Configure the physical layer on each up/downlink
* Implement & split off checkOutputPower

Moreover, the persistent buffers address calculations in for the Nonces & Session buffers was prone to user error - in fact, the session buffer even contained 48 useless bytes because I made an error! The improved implementation isn't particularly easier to read, but sure is safer. (And as a bonus, reduces from 396 bytes to 348.)

* [LoRaWAN] Improve persistent buffers addresses

... and the other bugfixes / changes:

* [LoRaWAN] Add getter for Time-on-Air of last uplink
* Update keywords.txt
* [LoRaWAN] Only allow sending each MAC request once per uplink (and only if piggybacking space is available)
* [LoRaWAN] Prevent possible underflows / bad delays while handling downlink windows
* [LoRaWAN] Remove unnecessary dynamic array

### CODE REVIEW

**Comment by StevenCellist** [comment]
This PR has been tested on EU868/OTAA using the Reference and Persistent sketches. Both feedback and additional tests are welcome.

File: src/modules/SX127x/SX1278.cpp [hunk 1]
```diff
@@ -342,6 +335,29 @@ int16_t SX1278::setOutputPower(int8_t power, bool useRfo) {
...
+      *clipped = RADIOLIB_MAX(-3, RADIOLIB_MIN(15, power));
+    }
+    RADIOLIB_CHECK_RANGE(power, -3, 15, RADIOLIB_ERR_INVALID_OUTPUT_POWER);
+  } else {
+    // PA_BOOST output, check high-power operation
+    if(clipped) {
+      if(power == 20) {
+        *clipped = 20;
+      }
+      *clipped = RADIOLIB_MAX(2, RADIOLIB_MIN(17, power));
```
Comment (jgromes) [inline src/modules/SX127x/SX1278.cpp:352]: Cppcheck (rightfully) complains about this one: https://github.com/jgromes/RadioLib/actions/runs/8884715530/job/24394437946?pr=1080#step:4:2427

**Review by jgromes — CHANGES_REQUESTED** [review]
Thank you @StevenCellist for another great contribution! I left some comments, but I think most of it can be addressed quite quickly.

File: src/modules/CC1101/CC1101.h [hunk 2]
```diff
@@ -774,6 +774,15 @@ class CC1101: public PhysicalLayer {
...
     int16_t setOutputPower(int8_t pwr);
 
+    /*!
+      \brief Check if output power is configurable.
+      \param power Output power in dBm.
+      \param clipped Clipped output power value to what is possible within the module's range.
+      \param raw Raw internal value.
+      \returns \ref status_codes
+    */
+    int16_t checkOutputPower(int8_t power, int8_t* clipped, uint8_t* raw);
```
[the remark this answers is not in the document]
Comment (StevenCellist) [reply src/modules/CC1101/CC1101.h:784]: Actually, I don't know exactly how this inheritance stuff works. I thought I saw something similar for some of the other methods so thought I'd copy it over, but apparently that's not how it works ;)
How should I solve this? (Or can you do that?)

File: src/protocols/LoRaWAN/LoRaWAN.cpp [hunk 3]
```diff
@@ -1987,7 +2005,8 @@ void LoRaWANNode::setDwellTime(bool enable, RadioLibTime_t msPerUplink) {
 uint8_t LoRaWANNode::maxPayloadDwellTime() {
   // configure current datarate
   DataRate_t dr;
-  findDataRate(this->dataRates[RADIOLIB_LORAWAN_CHANNEL_DIR_UPLINK], &dr);
+  // TODO this may fail horribly?
+  (void)findDataRate(this->dataRates[RADIOLIB_LORAWAN_CHANNEL_DIR_UPLINK], &dr);
```
[the remark this answers is not in the document]
Comment (StevenCellist) [reply src/protocols/LoRaWAN/LoRaWAN.cpp:2009]: This method is problematic for modules that do not support SF7-SF12, as in some weird way maybe a datarate is going to be configured that is not supported by the module. The function does not account for this (yet). But low priority so leaving it as a TODO for now.

File: src/protocols/LoRaWAN/LoRaWAN.cpp [hunk 4]
```diff
@@ -2224,25 +2227,20 @@ bool LoRaWANNode::execMacCommand(LoRaWANMacCommand_t* cmd) {
...
-        state = RADIOLIB_ERR_INVALID_OUTPUT_POWER;
-        while(state == RADIOLIB_ERR_INVALID_OUTPUT_POWER) {
-          // go from the highest power and lower it until we hit one supported by the module
-          state = this->phyLayer->setOutputPower(pwr--);
-        }
-        // only acknowledge if the requested datarate was succesfully configured
-        if(state == RADIOLIB_ERR_NONE) {
+        int8_t power = this->txPowerMax - 2*txSteps;
+        int8_t powerActual = 0;
+        state = this->phyLayer->checkOutputPower(power, &powerActual);
```
[the remark this answers is not in the document]
Comment (StevenCellist) [reply src/protocols/LoRaWAN/LoRaWAN.cpp:2233]: Same point as CC1101 with the `powerRaw` argument

File: src/modules/CC1101/CC1101.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (jgromes) [reply src/modules/CC1101/CC1101.h:784]: It works by haveing an exact match in the method signature (and preferrably, marking the overriding method as `override`). This also leads to the fact that you cannot have and overriding method which has some extra arguments, even if you do provide default values for them. An example is in `SX126x`:

```
/*!
      \brief Sets the module to standby mode (overload for PhysicalLayer compatibility, uses 13 MHz RC oscillator).
      \returns \ref status_codes
    */
    int16_t standby() override;

    /*!
      \brief Sets the module to standby mode.
      \param mode Oscillator to be used in standby mode. Can be set to RADIOLIB_SX126X_STANDBY_RC (13 MHz RC oscillator)
      or RADIOLIB_SX126X_STANDBY_XOSC (32 MHz external crystal oscillator).
      \param wakeup Whether to force the module to wake up. Setting to true will immediately attempt to wake up the module.
      \returns \ref status_codes
    */
    int16_t standby(uint8_t mode, bool wakeup = true);
```

Having just a single `standby(uint8_t mode = RADIOLIB_SX126X_STANDBY_XOSC, bool wakeup = true)` would not be able to override `PhysicalLayer::standby(void)`, that's why an additional `SX126x::standby(void)` is needed.

I will fix it in this branch, my guess is that you tested with SX126x which does not have this problem with `checkOutputPower`.

**Review by HeadBoffin — APPROVED** [review]
Random approval because Steven said so

**Comment by StevenCellist** [comment]
@jgromes thank you, LGTM. As per @HeadBoffin: tested on US915/OTAA as well.

**Review by jgromes — APPROVED** [review]

**Comment by jgromes** [comment]
@StevenCellist I pushed the changes and also tested using an SX1272 to make sure the overrides are OK so this should be good to go. Thanks!

**Pull request merged**